### PR TITLE
Bump maintenance-package dependency versions to latest conditioned to source build only

### PR DESCRIPTION
**Pull request by carlossanlop**
Similar to https://github.com/dotnet/roslyn/pull/76076

Needed to unblock https://github.com/dotnet/sdk/pull/45042

Should solve error:
```
/vmr/src/msbuild/src/Utilities/Microsoft.Build.Utilities.csproj(0,0): error NU1109: (NETCORE_ENGINEERING_TELEMETRY=Restore) Detected package downgrade: System.Memory from 4.6.0 to centrally defined 4.5.5. Update the centrally managed package version to a higher version. 
 Microsoft.Build.Utilities.Core -> System.Collections.Immutable 10.0.0-alpha.1.24603.1 -> System.Memory (>= 4.6.0) 
 Microsoft.Build.Utilities.Core -> System.Memory (>= 4.5.5)
```

Will also create a patch containing these changes for the sdk PR.

### CODE REVIEW

File: eng/Versions.props [hunk 1]
```diff
@@ -20,23 +20,35 @@
     <UsingToolVSSDK>true</UsingToolVSSDK>
```
Comment (carlossanlop) [inline eng/Versions.props:20]: GitHub did a poor job at showing the diff. I suggest using "split view" to understand how and which lines were moved.

**Review by MichaelSimons — APPROVED** [review]

**Comment by carlossanlop** [comment]
@MichaelSimons judging by the error, I suspect we need to manually trigger an SBRP deps flow, since those package versions have already been added to SBRP. I could include it in this PR.

```
.packages/microsoft.dotnet.arcade.sdk/9.0.0-beta.24562.13/tools/SourceBuild/AfterSourceBuild.proj(81,5): error : (NETCORE_ENGINEERING_TELEMETRY=AfterSourceBuild) 4 new pre-builts discovered! Detailed usage report can be found at /__w/1/s/artifacts/sb/prebuilt-report/baseline-comparison.xml.
See https://aka.ms/dotnet/prebuilts for guidance on what pre-builts are and how to eliminate them.
Package IDs are:
System.Buffers.4.6.0
System.Memory.4.6.0
System.Numerics.Vectors.4.6.0
System.Runtime.CompilerServices.Unsafe.6.1.0
```

**Comment by carlossanlop** [comment]
msbuild is consuming the source-build-reference-packages subscription from release/9.0, not from main. I should backport them.

File: eng/Versions.props [hunk 2]
```diff
@@ -20,23 +20,35 @@
...
   </PropertyGroup>
   <!-- Production Dependencies -->
-  <PropertyGroup>
+  <!-- Condition consumption of maintenance-packages dependencies based on source build.
+       This is to prevent "package downgrade" errors coming from other packages that are
+       already consuming the newest version of these same dependencies. -->
+  <PropertyGroup Condition="'$(DotNetBuildSourceOnly)' == 'true'">
+    <!-- Use newest package versions. -->
+    <MicrosoftIORedistVersion>6.1.0</MicrosoftIORedistVersion>
```
Comment (MichaelSimons) [inline eng/Versions.props:28]: It looks like this is not actually referenced in the SB configuration.  See https://github.com/dotnet/msbuild/blob/main/src/Directory.BeforeCommon.targets#L113

Comment (carlossanlop) [reply eng/Versions.props:28]: Ah good catch. I can revert this line change. Thanks!

**Comment by carlossanlop** [comment]
This is currently blocked by https://github.com/dotnet/source-build-reference-packages/pull/1091

File: eng/Versions.props (hunk 2 above)
Comment (carlossanlop) [reply eng/Versions.props]: Fixed.

**Comment by akoeplinger** [comment]
https://github.com/dotnet/msbuild/pull/11081 is bringing in the new source-build-reference-packages version

**Review by rainersigwald — APPROVED** [review]

**Comment by carlossanlop** [comment]
All green, @rainersigwald . Can you please merge?

**Review by YuliiaKovalova — APPROVED** [review]

**Comment by carlossanlop** [comment]
> All green, @rainersigwald . Can you please merge?

Or anyone from @dotnet/kitten please. 🙂

**Pull request merged**